Faster `Vector` concatenation

ansvonwa opened the pull request:
So far, concatenating two Vectors `v1` and `v2` of lengths $m$ (`v1`) and $n$ (`v2`) took $O(n)$ time.

This PR reduces the time needed to $O(\min(m, n))$.

For $m\lt n$, instead of pre-filling the `VectorBuilder` with `v1` and traversing all of `v2`, we now set an offset, such that after traversing `v1` and adding it's elements, the structure of the Builder does exactly match `v2`'s structure. This way, we can reuse arrays of `v2` and don't need to iterate over all of `v2`.

As an example, take the brown ("old misaligned") and red ("misaligned") line in the following plot. It shows times taken by concatenation of two Vectors with 1 million entries in total. The percentage states the relative size of `v1`. (So, $20\\%$ means $m=2\times 10^5, n=8\times10^5$.)
The values are in $\frac{\text{ns}}{\text{operation}}$. See also [benchmark source](https://github.com/ansvonwa/scala/blob/vector-faster-prepend/test/benchmarks/src/main/scala/scala/collection/immutable/VectorConcatBigVectorsBenchmark.scala).

The brown ("old misaligned") line shows the time for worst case (which is the case in 31 of 32 cases) of the old implementation.
The red ("misaligned") line shows times with this optimization. You see the improvement for $m\lt n$.

![s c i Vector concat](https://user-images.githubusercontent.com/9501177/192121750-9bd3795a-5625-47aa-ace4-08f401671269.png)

The way in which this optimization is implemented, yields another improvement. In special cases, time goes down to $O(\log_{32}(m+n))$, so basically zero.

In the old implementation, if the data arrays of the Vectors align (i.e. `(v1.suffix1.length + v2.prefix1.length) % 32 == 0`), there was only one call to `System.arraycopy` for each data array in `v2`, instead of two.
Thus, there was a small performance improvement for aligned Vectors (green and pale blue).

Now, we check for every slice whether it aligns to the Builder's structure. If that is the case, we reuse the old Array instead of copying it's content. We do this on every level. Thus, if the Vectors align at the lowest level (`(v1.suffix1.length + v2.prefix1.length) % WIDTH == 0`; $\frac1{32}$ chance for random vectors), we get the improvement shown by the yellow line. If the Vectors align at all levels (`(v1.suffix1…X.length + v2.prefix1…X.length) % WIDTHX == 0`), we get the dark blue times, so just a hand full of operations.

----

In short, the four commits do
1. fix old issues, mostly related to large Vectors,
2. detect where slices align and reuse arrays,
3. add a method (`alignTo`) to set the offset and another one to drop leading `null`s in Arrays before constructing the result,
4. change bit-arithmetics to a switch with hardcoded values.
   This should improve performance slightly, but I did not really get significantly shorter times in the benchmarks.
   It's mostly a matter of taste, but with the switch, it looks and feels more like the rest of the Vector class.


closes scala/bug#4442

ansvonwa committed:
expand bit-manipulations to switch in Vector.addArrN

ansvonwa committed:
fix minor issues concerning large Vectors before implementing faster concatenation

* misleading indentation
* `a6 = new Arr6(LASTWIDTH)`:
  The whole point of LASTWIDTH is that Arr6 may be larger than the other arrays.
  This is considered in all initializations except here.
  The added test shows how initFrom failed so far on arrays with size in
  (Int.MaxValue/2, Int.MaxValue] and works now.
* `else if (xor < WIDTH6) { // level = 5`:
  Similarly to the last one, Arr6 may have a size up to LASTWIDTH, so this check
  fails for valid Arr6s with size in [32,64]. About WIDTH6, see next point.
* `BITS6` and `WIDTH6`:
  There is no valid usage for them. The BITSn and WIDHTn variables mark which bits
  of the index correspond to which level, or more precisely Vectors up to level n
  have sizes of [0, WIDTHn). Thus, the correct value of BITS6 would be (BITS * 6 + 1),
  thus WIDTH6 would be Int.MaxValue+1, one "more" than the longest possible vector
  of level 6.
  The current value is just wrong, as it arbitrarily "cuts" the range of 6 bits
  which compose the index in Arr6 in 1 high and 5 lower bits.
  In fact, the line above was the only place where one of `BITS6` and `WIDTH6` was used.
  I suspect this was a little oversight because of the irregularity of level 6.

ansvonwa committed:
reuse Arrays in VectorBuilder.addVector(..) where possible

ansvonwa committed:
add VectorBuilder.alignTo and fast Vector.prependedAll

* add alignTo and leftAlignPrefix in VectorBuilder:
  alignTo(n: Int, v: Vector) ensures that if there are added n elements
  before Vector v, the Builder is aligned to the structure of v.
  This way, the Arrays of v can be reused in the Builder, which improves
  performance drastically.
  Formally, alignTo sets `len1`, `lenRest` and the internal arrays as
  if there were $x = len1 + lenRest$ elements already added, such that
  $x + n + prefixLenght(v)$ is a  multiple of the maximum prefix length
  of a Vector of v's level.
  The number $x$ is stored in `offset` and the new variable
  `prefixIsRightAligned` is set true to record the fact that in each
  prefix array, the length is WIDTH and the content is right-aligned.
  This is different to the normal case, where the prefixes are either
  full and size WIDTH or, if `initFrom` was used, full and trimmed.
  To remove the leading null entries in the prefix vectors, a call to
  leftAlignPrefixes is needed. This is done at the beginning of
  result().
* prependedAll0 and appendedAll0:
  In case that an IterableOnce (left side; may be a Vector itself) and a
  Vector (right side) are concatenated and the left one is shorter
  than the right Vector, a VectorBuilder aligned to the right Vector is
  used.
  `++` and `concat` forward to `appendedAll`, so they profit from this
  optimization.
* add tests and benchmark

ansvonwa commented:
/rebuild
Afaict, this change has no influence on `test.scala.sys.process.ProcessTest.t10823(ProcessTest.scala:92)`, which failed.

som-snytt commented:
I think you're off the hook, so to speak. I created a ticket! I'm pretty sure it has failed before.

SethTisue commented:
Thank you — this looks like it will a fantastic addition to Scala 2.13.11! Finding reviewers might take a bit of time.

retronym commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -1559,6 +1566,108 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
+      val i = (offset >>> BITS4) & MASK
+      if (i > 0) {
+        a = copyOfRange(a, i, WIDTH)
+        if (depth == 5) a5 = a.asInstanceOf[Arr5]
```
I'm doing some "poor man's test coverage" checks by replacing expressions with `???`

The `true` case of this `if` is not covered by the unit tests, nor is the analagous line for other depths below. What data would be need to get here?

retronym replied:
Here's the spots I found that don't fail the unit test.

```diff
diff --git a/src/library/scala/collection/immutable/Vector.scala b/src/library/scala/collection/immutable/Vector.scala
index 7dd3334e7d..8c04416feb 100644
--- a/src/library/scala/collection/immutable/Vector.scala
+++ b/src/library/scala/collection/immutable/Vector.scala
@@ -1606,8 +1606,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
       a = a6.asInstanceOf[Array[AnyRef]]
       val i = offset >>> BITS5
       if (i > 0) {
-        a = copyOfRange(a, i, LASTWIDTH)
-        a6 = a.asInstanceOf[Arr6]
+        ???
       }
       aParent = a
       a = a(0).asInstanceOf[Array[AnyRef]]
@@ -1617,7 +1616,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
       val i = (offset >>> BITS4) & MASK
       if (i > 0) {
         a = copyOfRange(a, i, WIDTH)
-        if (depth == 5) a5 = a.asInstanceOf[Arr5]
+        if (depth == 5) ???
         else aParent(0) = a
       }
       aParent = a
@@ -1628,7 +1627,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
       val i = (offset >>> BITS3) & MASK
       if (i > 0) {
         a = copyOfRange(a, i, WIDTH)
-        if (depth == 4) a4 = a.asInstanceOf[Arr4]
+        if (depth == 4) ???
         else aParent(0) = a
       }
       aParent = a
@@ -1639,7 +1638,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
       val i = (offset >>> BITS2) & MASK
       if (i > 0) {
         a = copyOfRange(a, i, WIDTH)
-        if (depth == 3) a3 = a.asInstanceOf[Arr3]
+        if (depth == 3) ???
         else aParent(0) = a
       }
       aParent = a
@@ -1650,7 +1649,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
       val i = (offset >>> BITS) & MASK
       if (i > 0) {
         a = copyOfRange(a, i, WIDTH)
-        if (depth == 2) a2 = a.asInstanceOf[Arr2]
+        if (depth == 2) ???
         else aParent(0) = a
       }
       aParent = a
@@ -1661,7 +1660,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
       val i = offset & MASK
       if (i > 0) {
         a = copyOfRange(a, i, WIDTH)
-        if (depth == 1) a1 = a.asInstanceOf[Arr1]
+        if (depth == 1) ???
         else aParent(0) = a
       }
     }
@@ -1757,9 +1756,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
         }
       case 6 => // note width is now LASTWIDTH
         if (lenRest % WIDTH5 != 0) {
-          // lenRest is not multiple of WIDTH5, so this slice does not align, need to try lower dimensions
-          slice.foreach(e => addArrN(e.asInstanceOf[Array[AnyRef]], 5))
-          return
+          ???
         }
         val copy1   = mmin((BITS * 6 + 1 - lenRest) >>> BITS5, sl)
         val copy2   = sl - copy1
@@ -1767,8 +1764,7 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
         System.arraycopy(slice, 0, a6, destPos, copy1)
         advanceN(WIDTH5 * copy1)
         if (copy2 > 0) {
-          System.arraycopy(slice, copy1, a6, 0, copy2)
-          advanceN(WIDTH5 * copy2)
+          ???
         }
     }
   }
```

ansvonwa replied:
> What data would be need to get here?

Should require be something like:
* The builder is filled up to `a5` (but not `a6`)
* `offset` is so large that `a5` starts with a `null` entry

Off the top of my head, I don't think this should occur in normal operation, as in such a case, [the modulo in `alignTo`](https://github.com/scala/scala/blob/e4b872afd9d197daf9e73dca0fbd289e2cf49d71/src/library/scala/collection/immutable/Vector.scala#L1584) should have prevented that. The only case that comes to my mind is if somebody called `alignTo` (when it's eventually public) with a different vector than what they later add to the builder. That is not likely to make sense but isn't really forbidden either. Maybe something like

```scala
  val v6 = ... // Vector6
  val vb = new VectorBuilder()
  vb.alignTo(23, v6)
  ab.addOne(..) ...// add 23 values
  val v5 = dropSomethingFromTheRight(v6) // drops so much that we now have a Vector5, but the offset is still set
  vb.addAll(v5)
  vb.result()
```

Anyway, good catch! I'm gonna try to find test cases for all the uncovered branches.

ansvonwa replied:
As mentioned, the branches were supposed to handle "misleading" arguments to `alignTo`. Problem was that `result()` cannot handle the `null` entries this created. Thus, if necessary, I now move the entries to the left and lower `offset` (and `lenRest`) accordingly. In some cases this may even reduce the `depth`.
```scala
        shrinkOffsetIfToLarge(WIDTH4)
        if ((lenRest >>> BITS4) == 0) depth = 4
```
Also, I now use `System.arraycopy` to keep `a` at it's current size. This is necessary to allow reusing the Builder after a call to `result()`.
```scala
        if (i > 0) System.arraycopy(a, i, a, 0, WIDTH - i)
```

This should explain most of the diff in `leftAlignPrefix()`.

---

The second of the two other branches you found, was actually useless, I removed it.

---

Every branch you mentioned should be covered by tests now.

SethTisue commented:
re: variety in test data, how about adding something to `test/scalacheck`?

retronym reviewed (commented):
This looks amazing. Thanks for presenting it so clearly.

I can't think of any performance tradeoffs to consider, am I right in thinking this is all upside?

If so, the risk seems to be a data-depedent bug in the new implementation. I've highlighted a few branches that aren't covered by the `s.c.i.VectorTest` (although may well be covered by other broader tests, I didn't run the full suite.). Let's try to expand the unit test coverage to at least cover these.

I'm always a bit wary of overflow bugs, but I don't see anything obvious here. I understand @szeiger might be able to take a look at this change, his experience with this implementation would be more likely to spot that sort of problem.

retronym reviewed (changes requested):
Expand unit test to cover more branches of the new implementation.

SethTisue commented:
We also have generative testing in the https://github.com/scala/scala-collection-laws repository.

The Scala 2 community build runs scala-collection-laws against merged PRs, but we can also do runs ourselves locally against the unmerged PR, using the PR validation snapshot (as documented in the README here in scala/scala).

In my clone of the scala-collection-laws repo, just now I did:

```
set Global / resolvers += "pr" at "https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots/"
++2.13.9-bin-e4b872a-SNAPSHOT!
test
```

and it passed, so that's encouraging. We should do the check again just before merge.

SethTisue commented:
This might be a change where actually enabling scoverage would have good cost/benefit ratio.

ansvonwa commented:
> I can't think of any performance tradeoffs to consider, am I right in thinking this is all upside?

Mostly, yes. The only thing: As the calls to `alignTo` and `leftAlignPrefix` take some (small but mostly constant) time, one may want to choose this optimization only if `left.length < right.length + c` for `left ++ right` and a small constant `c`, instead of the current `left.length < right.length`. This is only relevant for rather small `Vector`s that are large enough to trigger the optimization. I haven't actually tested how much time `alignTo` and `leftAlignPrefix` take, maybe I should.

Another (more theoretical) downside I could think of is, that as aligned `Vector`s are now _much_ faster to concatenate: If one only tests with those, they're code may run much slower than expected at production time. But imho definitely no reason to slow these cases down ;)


> I'm always a bit wary of overflow bugs

Most should be catched by the [`<= 0` check in `advance1`](https://github.com/scala/scala/blob/e4b872afd9d197daf9e73dca0fbd289e2cf49d71/src/library/scala/collection/immutable/Vector.scala#L1819), but I could add tests for that as well.
And note that an overflow in the builder could already happen when constructing `Vector`s of size about `(Int.MaxValue - WIDTH5 + 2)` if the offset is large enough (i.e. `prefix` is not full). This also yields an [ioob](https://github.com/scala/scala/blob/e4b872afd9d197daf9e73dca0fbd289e2cf49d71/src/library/scala/collection/immutable/Vector.scala#L1868).

---

Something else: Even though this "optimizes Vector concatenation" (maybe asymptotically even as much as possible with the current homogeneous `Vector` structure), maybe scala/bug#4442 should stay open?
There is still the possibility to introduce some form of heterogeneous structure that allows $O(log(n))$ concatenation also for misaligned `Vector`s. I'm currently experimenting with this, and it looks promising, but it's definitely not possible without some tradeoffs in (random) access times, at least for large `Vector`s that were obtained by concatenating many misaligned medium-sized `Vectors`.

ansvonwa commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -1559,6 +1566,108 @@ final class VectorBuilder[A] extends ReusableBuilder[A, Vector[A]] {
+      a = a6.asInstanceOf[Array[AnyRef]]
+      val i = offset >>> BITS5
+      if (i > 0) {
+        a = copyOfRange(a, i, LASTWIDTH)
```
Note to myself: `copyOfRange` breaks VB, because after `result()` `aN.length` may be `< WIDTH` which is illegal

ansvonwa committed:
make VectorBuilder reusable again

* `leftAlignPrefix()` now keeps the length of the outermost array at
  (LAST)WIDTH.
  This is necessary because after a call to `result()`, the VB may be
  used further.
* Allow arbitrarily misleading/wrong arguments in `alignTo()`
  Should not fail now, even if the hint given is completely nonsensical,
  `leftAlignPrefix()` now compensates for that.
  Negative alignments are allowed, e.g. if you plan to drop elements of
  the vector to be added.
* Add a bunch of tests to cover previously uncovered branches.

retronym commented:
A few unused imports are failing the build on the latest commits.

```
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/test/junit/scala/collection/immutable/VectorTest.scala:9:49: Unused import
[error] import scala.collection.immutable.VectorInline.{WIDTH, WIDTH2, WIDTH3, WIDTH4, WIDTH5}
[error]                                                 ^
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/test/junit/scala/collection/immutable/VectorTest.scala:9:56: Unused import
[error] import scala.collection.immutable.VectorInline.{WIDTH, WIDTH2, WIDTH3, WIDTH4, WIDTH5}
[error]                                                        ^
[error] two errors found
```

SethTisue commented:
> A few unused imports are failing the build on the latest commits.

You can get the same errors locally with `set Global / fatalWarnings := true`

ansvonwa commented:
/rebuild
Thank you for the hints.
Now, the build fails even earlier while using sdkman ("Do you want java 8.0.352-tem to be set as default? (Y/n): No output has been received in the last 10m0s").
Locally, I get errors at `scala.reflect.io.ZipArchiveTest.manifest resources just works` and `scala.tools.nsc.classpath.ZipAndJarFileLookupFactoryTest.manifest classpath entry works` but I don't think that has to do with these changes.

som-snytt commented:
@SethTisue wanted to know how often the spec build fails with spurious red X.

@ansvonwa what is your local error? as it should just work. Maybe something different on Windows versus CI?

ansvonwa commented:
/rebuild
(Last try may have been too soon)

retronym commented:
The spec build seems to be failing because sdkman is asking for confirmation:

```
Installing: java 8.0.352-tem

Done installing!

Do you want java 8.0.352-tem to be set as default? (Y/n): 

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```


@SethTisue perhaps we need this sprinkled somewhere?

```
export sdkman_auto_answer=true 
```

https://sdkman.io/usage#config
https://github.com/sdkman/sdkman-cli/issues/682sdk

SethTisue commented:
I've opened a ticket on the spec build failure: https://github.com/scala/scala-dev/issues/824

ansvonwa committed:
use VB.alignTo only if RHS is significantly larger

* To cempensate for alignTo's (small, but constant) overhead, it is only
called when the right Vector is mor then 64 elemente longer than the
left one.
* replaced some `….knownSize` by `k`

ansvonwa commented:
In addition to the remaining tests:
* I made sure VectorBuilder is reusable again,
* Made benchmarks on what the best offset is to compensate for `alignTo`'s time (and came to 64 entries; So only if the RHS is 64 entries longer than the LHS, we use `alignTo`)
* The (previously good) optimization in prependedAll0 and [appendedAll0](https://github.com/scala/scala/blob/7b8c40771b43b53d6ce680508d5f4034c5de1bee/src/library/scala/collection/immutable/Vector.scala#L236) is not optimal anymore. This is only, because concatenation via VB got faster, so it shouldn't be much of a problem. Anyway, in the long term, that should be adjusted accordingly. I think, as there is no drawback compared to before, this could be done in a different PR?

retronym commented:
Sorry for the delay! This looks good to go, thanks for going the extra mile.

SethTisue commented:
@ansvonwa I gave this PR top billing in the Scala 2.13.11 release notes. also sent out some kudos at https://fosstodon.org/@SethTisue/110540603310328217 . thank you again for this valuable work.

ansvonwa commented:
@SethTisue Wow, thank you!
Btw there may be something similar coming within the next year or so :)
